packages/big-design/src/components/Tree/reducer.ts

```diff
@@ -0,0 +1,276 @@
+  | { type: 'FOCUS_LEFT'; id: TreeNodeId }
+  | { type: 'FOCUS_FIRST' }
+  | { type: 'FOCUS_LAST' }
+  | { type: 'SELECTED_NODE'; values: T[] };
+
+export const createReducer = <T>(): Reducer<TreeState<T>, Action<T>> => (state, action) => {
+  switch (action.type) {
+    case 'TOGGLE_NODE':
+      if (state.expandedNodeIds.includes(action.id)) {
+        return {
+          ...state,
+          expandedNodeIds: state.expandedNodeIds.filter((nodeId) => nodeId !== action.id),
+          visibleNodeIds: removeChildNodes(state.visibleNodeIds, action.id, state.nodeMap),
+        };
+      }
+
+      return {
+        ...state,
+        expandedNodeIds: insertNode(state.expandedNodeIds, action.id, state.flattenedNodeIds),
+        visibleNodeIds: buildVisibleNodes(state.nodeMap[action.id].children, state),
+      };
+
+    case 'FOCUS':
+      return {
+        ...state,
+        focusedNode: action.id,
+      };
+
+    case 'FOCUS_DOWN':
+      return {
+        ...state,
+        focusedNode: getNextVisibleNode(state, action.id),
+      };
+
+    case 'FOCUS_UP':
+      return {
+        ...state,
+        focusedNode: getPreviousVisibleNode(state, action.id),
+      };
+
+    case 'FOCUS_RIGHT':
+      if (state.expandedNodeIds.includes(action.id)) {
+        return {
+          ...state,
+          focusedNode: getNextVisibleNode(state, action.id),
+        };
+      }
+
+      if (state.nodeMap[action.id].children.length > 0) {
+        return {
+          ...state,
+          expandedNodeIds: insertNode(state.expandedNodeIds, action.id, state.flattenedNodeIds),
+          visibleNodeIds: buildVisibleNodes(state.nodeMap[action.id].children, state),
+        };
+      }
+
+      return state;
+
+    case 'FOCUS_LEFT':
+      if (state.expandedNodeIds.includes(action.id)) {
+        return {
+          ...state,
+          expandedNodeIds: state.expandedNodeIds.filter((nodeId) => nodeId !== action.id),
+          visibleNodeIds: removeChildNodes(state.visibleNodeIds, action.id, state.nodeMap),
+        };
+      }
+
+      if (state.nodeMap[action.id].parent !== undefined) {
+        return {
+          ...state,
+          focusedNode: state.nodeMap[action.id].parent,
+        };
+      }
+
+      return state;
+
+    case 'FOCUS_FIRST':
+      return {
+        ...state,
+        focusedNode: state.visibleNodeIds[0],
+      };
+
+    case 'FOCUS_LAST':
+      return {
+        ...state,
+        focusedNode: state.visibleNodeIds[state.visibleNodeIds.length - 1],
+      };
+
+    case 'SELECTED_NODE':
+      if (action.values) {
+        return {
+          ...state,
+          selectedValues: action.values,
+        };
+      }
+
+      return state;
+
+    default:
+      return state;
+  }
+};
+
+const initializeNodeMap = <T>(nodes: TreeProps<T>['nodes'], parent?: TreeNodeId): NodeMap =>
+  nodes.reduce((acc, node) => {
+    const childNodeMap = node.children ? initializeNodeMap(node.children, node.id) : {};
```

Comment:
I notice that `initializeNodeMap`, `initializeExpandedNodeIds`, `initializeFlattenedNodeIds`, `initializeVisibleNodeIds`, `initializeSelectedValues` and `initializeFocusNode` functions traverse the same tree of `nodes` multiple times. Do you think if we can achieve the same result by traversing the tree only once?

Also I notice that the traversal is not tail-call optimised. It shouldn't be a problem unless we're dealing with a large list. Otherwise, we might have a memory and call-stack problem. So it'd be good to look at either optimising the recursive calls or use a while loop instead.

Reply:
I've been thinking about that since I initially pushed this PR. I'm going to modify the underlying state data structure to use `Set` and `Map` for most of the state. While I'll do that, I'll try to build the state through 1 traversal.

Regarding tail-call optimization, I'll take a look to see what I can do.

Reply:
After refactoring the state to use `Set` and `Map` I think I resolved this. Take a look and let me know what you think.